comfy/samplers.py

```diff
@@ -301,6 +315,47 @@ def blank_inpaint_image_like(latent_image):
     blank_image[:,3] *= 0.1380
     return blank_image
 
+def resolve_cond_masks(conditions, h, w, device):
+    # We need to decide on an area outside the sampling loop in order to properly generate opposite areas of equal sizes.
+    # While we're doing this, we can also resolve the mask device and scaling for performance reasons
+    for i in range(len(conditions)):
+        c = conditions[i]
+        if 'mask' in c[1]:
+            mask = c[1]['mask']
+            mask = mask.to(device=device)
+            modified = c[1].copy()
+            if len(mask.shape) == 2:
+                mask = mask.unsqueeze(0)
+            if mask.shape[2] != h or mask.shape[3] != w:
+                mask = torch.nn.functional.interpolate(mask.unsqueeze(1), size=(h, w), mode='bilinear', align_corners=False).squeeze(1)
+
+            if 'area' not in modified:
+                bounds = torch.max(torch.abs(mask),dim=0).values.unsqueeze(0)
+                if torch.max(bounds) == 0:
+                    # Handle the edge-case of an all black mask (where masks_to_boxes would error)
+                    area = (0, 0, 0, 0)
+                else:
+                    box = masks_to_boxes(bounds)[0].type(torch.int)
+                    H, W, Y, X = (box[3] - box[1] + 1, box[2] - box[0] + 1, box[1], box[0])
+                    # Make sure the height and width are divisible by 8
+                    if X % 8 != 0:
+                        newx = X // 8 * 8
+                        W = W + (X - newx)
+                        X = newx
+                    if Y % 8 != 0:
+                        newy = Y // 8 * 8
+                        H = H + (Y - newy)
+                        Y = newy
+                    if H % 8 != 0:
+                        H = H + (8 - (H % 8))
+                    if W % 8 != 0:
+                        W = W + (8 - (W % 8))
+                    area = (int(H), int(W), int(Y), (X))
+                modified['area'] = area
```

Comment:
+1 agreed.

Reply:
Just to make sure I understand, are you recommending sampling across the entire image even though only the area within the mask is modified?
Anecdotally, I got better results by limiting to the bounding box. What do you think of making a "auto_set_area" option on the Condition By Mask node (that defaults to off) to allow both behaviors?
(Using the Condition Set Area node in serial works well if the mask is pre-generated and the user can manually find bounds, but there isn't a clean way to handle dynamically generated masks with that node.)

Reply:
Yes because that's the behaviour I would expect: The entire image/set area is sampled but only the masked part is changed. This is how inpainting works too.

Allowing both behaviours is fine. I think there's a better way to do.

Another thing: using masks_to_boxes and depending on torchvision probably isn't the best or most efficient way to calculate the smallest area dimensions.

Reply:
Cool, sounds like a plan to me. Thanks!

Really? I didn't look into the implementation of masks_to_boxes -- I assumed that since it was included in torch (and all it does is calculate bounding boxes) it would be hyper-optimized for that task. What method would you suggest?

Reply:
Oh I thought it generated bounding boxes for all submasks, nevermind then.

Can you just include this code in then: https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py#L407

It's only a few lines long and we don't really depend on torchvision.